### PR TITLE
Performance Improvement

### PR DESCRIPTION
**Pull request by cedricpinson**

### CODE REVIEW

File: sources/osg/State.js [hunk 1]
```diff
@@ -1157,48 +1150,142 @@ State.prototype = MACROUTILS.objectLibraryClass( MACROUTILS.objectInherit( Objec
...
+        if ( internalArray.length < 16 ) {
+            if ( internalArray instanceof Float32Array ) {
+                cacheData = new Float32Array( internalArray.length );
+            } else if ( internalArray instanceof Int32Array ) {
+                cacheData = new Int32Array( internalArray.length );
+            } else if ( internalArray instanceof Array ) {
+                cacheData = new Array( internalArray.length );
+            } else {
+
+                console.log( 'not found' );
```
Comment (jtorresfabra) [inline sources/osg/State.js]: Notify.log?

**Comment by cedricpinson** [comment]
performance branch
screen shot 2017-04-05 at 18 46 51

master branch
screen shot 2017-04-05 at 16 52 29

**Comment by stephomi** [comment]
On the performance example without complexStateSet=true the performance branch is a lot slower for me (items 4 and depth 3).

From left to right : for-in-experiment, master, performance-test
Test with multiple loading and it's consistent

sans titre

**Comment by cedricpinson** [comment]
I merged the options for complex stateSet in master

File: sources/osg/Utils.js [hunk 2]
```diff
@@ -167,6 +167,117 @@ Utils.createPrototypeClass = function ( Constructor, prototype, libraryName, cla
...
+
+// ===============================================
+
+
+var typeMemberIndex = 0;
+var textureTypeMemberIndex = 0;
+var stateAttributeTypeMember = {};
+var textureStateAttributeTypeMember = {};
+var attributeTypeIndex = 0;
+var stateAttributeType = {};
```
Comment (jtorresfabra) [inline sources/osg/Utils.js]: Not sure is better to use maps, to avoid duplication

File: sources/osg/CullVisitor.js [hunk 3]
```diff
@@ -76,8 +76,19 @@ var CullVisitor = function () {
 
 };
 
+CullVisitor.registerApplyNodeType = function ( type, apply ) {
```
Comment (cedricpinson) [inline sources/osg/CullVisitor.js]: agree

File: sources/osg/CullVisitor.js [hunk 4]
```diff
@@ -477,7 +488,7 @@ CullVisitor.prototype = MACROUTILS.objectInherit( CullStack.prototype, MACROUTIL
 // Camera cull visitor call
 // ANY CHANGE, any change : double check in rendere Camera code
 // for the first camera
-CullVisitor.prototype[ Camera.typeID ] = function ( camera ) {
+var CameraApply = function ( camera ) {
```
Comment (cedricpinson) [inline sources/osg/CullVisitor.js]: agree

File: sources/osg/Utils.js (hunk 2 above)
Comment (cedricpinson) [inline sources/osg/Utils.js]: I dont understand {} there is no risk of duplication ?

Comment (jtorresfabra) [inline sources/osg/Utils.js]: forget it.

**Comment by Kuranes** [comment]
👍 Love it.
++ utils for osg object creation
++ cullVisitorHelper.registerApplyFunction 

in the helper in utils to create prototype of classes, could even had some assert (deveblock) checking for missing mandatory methods. ( for StateAttribute I would the "clone" which I had problem with each time I made a new stateattribute )

**Comment by cedricpinson** [comment]
@Kuranes Yeah that would be great to have cloneType also built by default in this helper, but we have some class that needs to do something specific  like:
```
    cloneType: function () {
        return new ShadowCastAttribute( true );
    },
```
It's not the only one, but it means we should change the default behavior to be sure default Constructor generate a default state. It's normal that in some case it needs something specific, but we will check that later

**Comment by jtorresfabra** [comment]
Needs rebase now :(

**Pull request merged**